server/src/main/java/org/opensearch/index/mapper/SemanticVersionFieldMapper.java

```diff
@@ -0,0 +1,313 @@
+            if (method == null) {
+                method = MultiTermQuery.CONSTANT_SCORE_REWRITE; // Default rewrite method
+            }
+            return new PrefixQuery(new Term(name(), indexedValueForSearch(value)), method);
+        }
+
+        @Override
+        public Query fuzzyQuery(
+            Object value,
+            Fuzziness fuzziness,
+            int prefixLength,
+            int maxExpansions,
+            boolean transpositions,
+            MultiTermQuery.RewriteMethod method,
+            QueryShardContext context
+        ) {
+
+            if (method == null) {
+                method = MultiTermQuery.CONSTANT_SCORE_REWRITE;
+            }
+
+            return new FuzzyQuery(
+                new Term(name(), indexedValueForSearch(value)),
+                fuzziness.asDistance(),
+                prefixLength,
+                maxExpansions,
+                transpositions,
+                method
+            );
+        }
+
+        @Override
+        public ValueFetcher valueFetcher(QueryShardContext context, SearchLookup searchLookup, String format) {
+            return new SourceValueFetcher(name(), context, format) {
+                @Override
+                protected String parseSourceValue(Object value) {
+                    return value.toString();
+                }
+            };
+        }
+
+        @Override
+        public IndexFieldData.Builder fielddataBuilder(String fullyQualifiedIndexName, Supplier<SearchLookup> searchLookup) {
+            return new SortedSetOrdinalsIndexFieldData.Builder(name() + "._normalized", CoreValuesSourceType.BYTES);
+        }
+
+        @Override
+        public Map<String, String> meta() {
+            return meta;
+        }
+    }
+
+    @Override
+    protected void parseCreateField(ParseContext context) throws IOException {
+        String value = context.parser().textOrNull();
+        if (value == null) {
+            return;
+        }
+
+        SemanticVersion version = SemanticVersion.parse(value);
+        String versionString = version.toString();
+        String normalizedValue = version.getNormalizedComparableString();
+        BytesRef bytes = new BytesRef(versionString);
+        BytesRef normalizedValueBytes = new BytesRef(normalizedValue);
+
+        // For retrieval: store original version string
+        context.doc().add(new StoredField(fieldType().name(), versionString));
+
+        // For searching (term queries): use original version string
+        context.doc().add(new KeywordField(fieldType().name(), bytes, Field.Store.YES));
+
+        // For range queries and sorting: use normalized form
+        context.doc().add(new KeywordField(fieldType().name() + NORMALIZED_FIELD_SUFFIX, normalizedValueBytes, Field.Store.NO));
+        context.doc().add(new SortedSetDocValuesField(fieldType().name() + NORMALIZED_FIELD_SUFFIX, normalizedValueBytes));
```

Comment:
Once we add the `hasDocValues` we would need to check if they are enabled before we store them